n-api: re-implement async env cleanup hooks

gabrielschulhof opened the pull request:
* Avoid passing core `void*` and function pointers into add-on.
* Document `napi_async_cleanup_hook_handle` type.
* Render receipt of the handle mandatory. Removal of the handle remains
  mandatory.

Fixes: https://github.com/nodejs/node/issues/34715
Signed-off-by: Gabriel Schulhof <gabriel.schulhof@intel.com>

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

gabrielschulhof commented on doc/api/n-api.md:
````diff
@@ -1612,13 +1650,15 @@ added: REPLACEME
 
 ```c
 NAPI_EXTERN napi_status napi_remove_async_cleanup_hook(
-    napi_env env,
````
I removed the need to pass `napi_env` here because we wish to discourage JS execution in environment cleanup hooks and this way it should prevent add-ons from having to store the `env` for the sole purpose of calling this function from a `uv_close` callback.

We store `napi_env` internally and it's guaranteed to be usable for the lifetime of a `remove_handle`, because we `Ref()` it during its construction and `Unref()` it asynchronously during its destruction.

addaleax commented on src/node_api.cc:
```diff
@@ -519,41 +519,65 @@ napi_status napi_remove_env_cleanup_hook(napi_env env,
+      env_(env),
+      user_hook_(user_hook),
+      user_data_(user_data) {
+    auto handle = node::AddEnvironmentCleanupHook(env->isolate, Hook, this);
+    handle_ = std::move(handle);
```
```suggestion
    handle_ = node::AddEnvironmentCleanupHook(env->isolate, Hook, this);
```

addaleax commented on src/node_api.cc:
```diff
@@ -519,41 +519,65 @@ napi_status napi_remove_env_cleanup_hook(napi_env env,
+
+  ~napi_async_cleanup_hook_handle__() {
+    node::RemoveEnvironmentCleanupHook(std::move(handle_));
+    if (done_cb_)
```
```suggestion
    if (done_cb_ != nullptr)
```

addaleax commented on doc/api/n-api.md:
```diff
@@ -1585,23 +1616,30 @@ added: REPLACEME
 NAPI_EXTERN napi_status napi_add_async_cleanup_hook(
     napi_env env,
-    void (*fun)(void* arg, void(* cb)(void*), void* cbarg),
+    napi_async_cleanup_hook hook,
```
If you're removing the signature here, please add a link to the definition of `napi_async_cleanup_hook`

gabrielschulhof replied:
Added the link [below](https://github.com/nodejs/node/pull/34819/files#diff-6782512949f60a32853d80270ceefeb0R5812).

mhdawson replied:
I do like how this better hides the internal implementation. It seemed like we were leaking a bit more than needed in the N-API APIs wrapping the internal node api.

gabrielschulhof commented:
@addaleax I have addressed your review comments.

Qualitatively, I think this approach makes `napi_add_async_cleanup_hook()` and `napi_remove_async_cleanup_hook` more into bookends for a specific cleanup operation. Furthermore, there is a nice symmetry produced by the fact that the entirety of the setup for a hook happens in its constructor, and the entirety of its teardown happens in its destructor. Additionally, it leaves only one value for the add-on maintainer to track.

I understand that these are all qualitative arguments and, as such, more or less a matter of taste. Quantitatively, perhaps, we can look at the test and see that it is simplified by 18 lines, though, since the change removes support for optional storage of the handle, the six removed lines that tested for that feature should perhaps not be considered. That still leaves 12 lines that are an indication that perhaps using async cleanup hooks by way of this API rather than the previous API results in less add-on code that needs writing and maintaining.

addaleax commented:
@gabrielschulhof I mean, I think any high-level wrapper like the C++ one would expose these two versions the same way anyway :)

Ultimately, I'm good with this if you think it's worth the breakage.

gabrielschulhof commented:
This is an overview for those who might wish to review of the current implementation vs. the proposed implementation showing how each would handle the most common use case:

```
Current implementation                               Proposed implementation
|                                                    |
v                                                    v
+------------------------------------------------+   +----------------------------------------------+
| Create async thing (e.g. uv handle)            |   | Create async thing (e.g. uv handle)          |
| napi_add_async_cleanup_hook(hook_cb)           |   | napi_add_async_cleanup_hook(hook_cb)         |
| Optional: Store returned hook_handle           | ! | Mandatory: Store returned hook_handle        |
+------------------------------------------------+   +----------------------------------------------+
|                                                    |
v                                                    v
+- hook_cb: -------------------------------------+   +- hook_cb: -----------------------------------+
| store done_cb (function ptr)                   | ! |                                              |
| store done_arg (void*)                         | ! |                                              |
| if a handle was stored                         | ! |                                              |
|   call napi_remove_async_cleanup_hook(handle)  | ! |                                              |
| Start closing the async thing (e.g uv_close)   |   | Start closing the async thing (e.g uv_close) |
+------------------------------------------------+   +----------------------------------------------+
|                                                    |
v                                                    v
+- uv_close cb: ---------------------------------+   +- uv_close cb: -------------------------------+
| call done_cb(done_arg)                         | ! | call napi_remove_async_cleanup_hook(handle)  |
+------------------------------------------------+   +----------------------------------------------+


+--+
|  | <-- an iteration of the event loop
+--+
```

gabrielschulhof commented:
@addaleax I added the `changes:` sections you requested.

Himself65 reviewed (approved):
LGTM

gabrielschulhof commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/32941/

mhdawson commented on src/node_api.cc:
```diff
@@ -519,41 +519,64 @@ napi_status napi_remove_env_cleanup_hook(napi_env env,
 
-  node::RemoveEnvironmentCleanupHook(std::move(remove_handle->handle));
-  delete remove_handle;
+  if (remove_handle == nullptr)
```
I think it can still be optional to get the handle when the hook is added. When the hook is called it is passed the handle so unless it wants to call remove before the hook runs, it does not need to get/store the handle.  I'd suggest we change back to this being optional.

mhdawson commented:
The assumption is that there should be little to no usage of the existing API as it's only be out for ~ 3weeks from what Gabriel tells me and as an Experimental API would should be able to change it in this case.

gabrielschulhof commented:
@mhdawson it is once more optional to receive and store the handle during creation. It is nevertheless still mandatory to store the handle when receiving it in the cleanup hook, because it must be passed to `napi_remove_async_cleanup_hook` when the handle cleanup completes:

```
Current implementation                               Proposed implementation
|                                                    |
v                                                    v
+------------------------------------------------+   +----------------------------------------------+
| Create async thing (e.g. uv handle)            |   | Create async thing (e.g. uv handle)          |
| napi_add_async_cleanup_hook(hook_cb)           |   | napi_add_async_cleanup_hook(hook_cb)         |
| Optional: Store returned hook_handle           |   | Optional: Store returned hook_handle         |
+------------------------------------------------+   +----------------------------------------------+
|                                                    |
v                                                    v
+- hook_cb: -------------------------------------+   +- hook_cb: -----------------------------------+
| store done_cb (function ptr)                   | ! |                                              |
| store done_arg (void*)                         | ! |                                              |
| if a handle was stored                         | ! | if a handle was not stored                   |
|   call napi_remove_async_cleanup_hook(handle)  | ! |   store it for use in the uv_close cb        |
| Start closing the async thing (e.g uv_close)   |   | Start closing the async thing (e.g uv_close) |
+------------------------------------------------+   +----------------------------------------------+
|                                                    |
v                                                    v
+- uv_close cb: ---------------------------------+   +- uv_close cb: -------------------------------+
| call done_cb(done_arg)                         | ! | call napi_remove_async_cleanup_hook(handle)  |
+------------------------------------------------+   +----------------------------------------------+


+--+
|  | <-- an iteration of the event loop
+--+
```

This actually reduces the breakage slightly because the first step is not identical to the current implementation.

gabrielschulhof commented:
Please mind the edit above, sorry:

`-` "...the first step is not identical..."
`+` "...the first step is now identical..."

gabrielschulhof commented:
Landed in 0848f56cb39432090cdb99af9b8541fbc1a2849c.
